### PR TITLE
fix: #1660 - correction for ping method to get proper status URL in case of a basepath provided

### PR DESCRIPTION
**Pull request by robinGupta11392**
## Change list

Fixed the logic of appending basepath to the Hub URL which fixes Issue #1660 - https://github.com/appium/java-client/issues/1660.

AppiumDriverLocalService was not able to start with Appium CLI versions < 2 as the ping() method was not able to get the status URL properly. Fixed the logic of appending basepath to sanitise the user input and appending basepath with "/" in the end, if not provided by user.
 
## Types of changes

What types of changes are you proposing/introducing to Java client?
_Put an `x` in the boxes that apply_

- [x] No changes in production code.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

Test Execution logs - https://gist.github.com/robinGupta11392/e8c37c7e10e03cd00a4d44d398402738
Issue fixed - https://github.com/appium/java-client/issues/1660
Changed Classes:
 - ServerBuilderTest
 - AppiumDriverLocalService
 
Method Added:
 - sanitizeBasePath
 
Method Modified:
 - withBasePath

Test Added:
 - checkAbilityToStartServiceUsingBasePath

### CODE REVIEW

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java [hunk 1]
```diff
@@ -94,10 +94,23 @@ public static AppiumDriverLocalService buildService(AppiumServiceBuilder builder
...
         return this;
     }
 
+    @SneakyThrows private static String sanitizeBasePath(String basePath) {
+        if (null == basePath) {
+            LOG.warn("Base Path cannot be NULL -- ignoring the basepath configuration");
+            return null;
+        }
+        basePath = basePath.trim();
+        if (basePath.isBlank() || basePath.isEmpty()) {
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:107]: `String#isBlank` is available starting from Java 11 only.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java [hunk 2]
```diff
@@ -94,10 +94,23 @@ public static AppiumDriverLocalService buildService(AppiumServiceBuilder builder
...
+            LOG.warn("Base Path cannot be Empty or Blank -- ignoring the basepath configuration");
+            return null;
+        }
+        return (basePath.endsWith("/") ? basePath : basePath + "/");
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:111]: redundant parenthesis
```suggestion
        return basePath.endsWith("/") ? basePath : basePath + "/";
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java [hunk 3]
```diff
@@ -94,10 +94,23 @@ public static AppiumDriverLocalService buildService(AppiumServiceBuilder builder
     }
 
     public AppiumDriverLocalService withBasePath(String basePath) {
-        this.basePath = basePath;
+        this.basePath = sanitizeBasePath(basePath);
         return this;
     }
 
+    @SneakyThrows private static String sanitizeBasePath(String basePath) {
+        if (null == basePath) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:102]: Consider using checkNotNull helper from guava lib. An exception must be thrown if this argument is set to null

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 1 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:107]: Apache commons lib also provides isBlank method

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java [hunk 4]
```diff
@@ -94,10 +94,23 @@ public static AppiumDriverLocalService buildService(AppiumServiceBuilder builder
     }
 
     public AppiumDriverLocalService withBasePath(String basePath) {
-        this.basePath = basePath;
+        this.basePath = sanitizeBasePath(basePath);
         return this;
     }
 
+    @SneakyThrows private static String sanitizeBasePath(String basePath) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:101]: what does this method throw?

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 3 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:102]: I believe the default value for basePath is null in case the user doesn't provide a basepath at all.. I referred to the getUrl method in which its checking for null. we should not throw an exception in this case (in my opinion) rather a warning message should be good enough and appium can start with no additional basepath

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 4 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:101]: I will remove the SneakyThrows, it came by mistakenly copy paste. thanks for highlighting.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 3 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:102]: But yeah, we can highlight to user by throwing an exception that they have provided a null value in basepath, so that they can make an informed decision to either provide a valid value or no value at all. I will make the change, thanks.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 1 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:107]: quick question on this - do we have any log/documentation which informs user that they need to provide a basepath in order to use Appium CLI v < v2? 
I believe we can pass a blank/empty string to it and it still works fine.

Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:107]: Just noticed that Blank String as argument values are not added to the appium command line while starting "main.js", will throw an exception if user tries to configure basepath as blank.

Below code which doesn't add value to the command line in case of blank

 ```
@Override
    protected ImmutableList<String> createArgs() {
...
Set<Map.Entry<String, String>> entries = serverArguments.entrySet();
        for (Map.Entry<String, String> entry : entries) {
...
if (!StringUtils.isBlank(value)) {
                argList.add(value);
            }
...
```

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 3 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:102]: the only place this API is called is `withBasePath` builder method. And there basePath must not be null. In general we are quite careful about explicitly returning null from any method and try to avoid that as much as possible.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 1 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:107]: > quick question on this - do we have any log/documentation which informs user that they need to provide a basepath in order to use Appium CLI v < v2?
> I believe we can pass a blank/empty string to it and it still works fine.

https://github.com/appium/java-client/blob/master/docs/v7-to-v8-migration-guide.md#appiumdriverlocalservice

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java [hunk 5]
```diff
@@ -94,10 +94,23 @@ public static AppiumDriverLocalService buildService(AppiumServiceBuilder builder
...
+        if (null == basePath) {
+            LOG.warn("Base Path cannot be NULL -- ignoring the basepath configuration");
+            return null;
+        }
+        basePath = basePath.trim();
+        if (basePath.isBlank() || basePath.isEmpty()) {
+            LOG.warn("Base Path cannot be Empty or Blank -- ignoring the basepath configuration");
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:108]: I usually avoid writing logs in the library code unless really necessary. They might be annoying and hard to disable.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 3 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java:102]: In general I'd add something like `checkArgument(isBlank(basePath), "basePath must not be empty or null")`

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 1 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java]: Fixed this in the new commit, please review.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 3 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java]: Fixed this by moving the validations to "withArgument" method and checking for null, blank and empty values. Introduced a InvalidBasePathException to throw in case of any validations fail.

File: src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java (hunk 5 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumDriverLocalService.java]: Removed the LOG in the new commit.

**Comment by robinGupta11392** [comment]
https://gist.github.com/robinGupta11392/19cb22749170119d791cbeb8e1a714b3 --> ServerBuilderTest execution logs with Appium V2 Beta 25 post the fix. It seems there is one Log test failing right from the moment i cloned it - maybe you guys can have a look for reference.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 6]
```diff
@@ -217,13 +218,25 @@ public AppiumServiceBuilder withArgument(ServerArgument argument, String value)
...
                 serverArguments.put(argName, value);
                 break;
         }
         return this;
     }
 
+    private static String sanitizeBasePath(String basePath) {
+        basePath = checkNotNull(basePath).trim();
+        if (StringUtils.isBlank(basePath) || basePath.isEmpty()) {
+            throw new InvalidBasePathException(
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:234]: IMHO it's an overkill to have a separate exception there. the standard InvalidArgumentException thrown by checkArgument would be totally fine

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 7]
```diff
@@ -217,13 +218,25 @@ public AppiumServiceBuilder withArgument(ServerArgument argument, String value)
...
             default:
                 serverArguments.put(argName, value);
                 break;
         }
         return this;
     }
 
+    private static String sanitizeBasePath(String basePath) {
+        basePath = checkNotNull(basePath).trim();
+        if (StringUtils.isBlank(basePath) || basePath.isEmpty()) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:233]: isBlank also returns true for empty strings

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 8]
```diff
@@ -312,4 +319,56 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
     }
+
+    @Test
+    public void checkAbilityToStartServiceUsingValidBasePathWithMultiplePathParams() {
+        String baseUrl = String.format("http://%s:%d/", BROADCAST_IP_ADDRESS, DEFAULT_APPIUM_PORT);
+        String basePath = "wd/hub";
+        service = new AppiumServiceBuilder().withArgument(BASEPATH, basePath).build();
+        service.start();
+        assertTrue(service.isRunning());
+        assertEquals(baseUrl + basePath + "/",service.getUrl().toString());
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:330]: Please format the stuff everywhere, so there are spaces after comma

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 9]
```diff
@@ -217,13 +218,25 @@ public AppiumServiceBuilder withArgument(ServerArgument argument, String value)
             case "-g":
                 withLogFile(new File(value));
                 break;
+            case "--base-path":
+                serverArguments.put(argName,sanitizeBasePath(value));
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:222]:
```suggestion
                serverArguments.put(argName, sanitizeBasePath(value));
```

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 7 above)
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:233]: no need to check for blank value after trim operation
```suggestion
        basePath = checkNotNull(basePath).trim();
        if (basePath.isEmpty()) {
```

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 10]
```diff
@@ -312,4 +319,56 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
+        assertTrue(service.isRunning());
+        assertEquals(baseUrl + basePath.substring(1) ,service.getUrl().toString());
+    }
+
+    @Test
+    public void checkAbilityToValidateBasePathForEmptyBasePath() {
+        try {
+            service = new AppiumServiceBuilder().withArgument(BASEPATH, "").build();
+            fail("Base path was not validated for Blank or Empty string");
+        } catch (Exception e) {
```
Comment (valfirst) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:348]: use `org.junit.Assert#assertThrows(java.lang.Class<T>, org.junit.function.ThrowingRunnable)` instead

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 9 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:222]: Noted on this, take care in the latest commit.

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 8 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:330]: Noted on this, take care in the latest commit.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 7 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:233]: Noted on this, take care in the latest commit.

Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:233]: Noted on this, taken care in the latest commit.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 6 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:234]: Noted on this, taken care in the latest commit.

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 11]
```diff
@@ -54,6 +57,9 @@
     private static final Path PATH_T0_TEST_MAIN_JS = ROOT_TEST_PATH
             .resolve("service").resolve("local").resolve("main.js");
 
+    private static final String INVALID_BASE_PATH_ERROR_MESSAGE =
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:60]: it is usually ok to just verify the exception class. Tests become too flaky because string messages could be changed too often

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 12]
```diff
@@ -54,6 +40,17 @@
...
+import lombok.SneakyThrows;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.SystemUtils;
+import org.apache.commons.validator.routines.InetAddressValidator;
+import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.Platform;
+import org.openqa.selenium.os.ExecutableFinder;
+import org.openqa.selenium.remote.Browser;
+import org.openqa.selenium.remote.service.DriverService;
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:53]: a lot of unnecessary changes in the import order

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 13]
```diff
@@ -217,13 +214,25 @@ public AppiumServiceBuilder withArgument(ServerArgument argument, String value)
...
         }
         return this;
     }
 
+    private static String sanitizeBasePath(String basePath) {
+        basePath = checkNotNull(basePath).trim();
+        if (basePath.isEmpty()) {
+            throw new IllegalArgumentException(
+                "Given base path is not valid - blank or empty values are not allowed for base path");
+        }
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:232]:
```suggestion
        Preconditions.checkArgument(!basePath.isEmpty(), "Given base path is not valid - blank or empty values are not allowed for base path");
```

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 14]
```diff
@@ -18,22 +21,22 @@
...
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.OutputStream;
 import java.nio.file.Path;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.List;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
```
Comment (valfirst) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:39]: unnecessary changes in the imports order

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 15]
```diff
@@ -312,4 +318,47 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
+        service = new AppiumServiceBuilder().withArgument(BASEPATH, basePath).build();
+        service.start();
+        assertTrue(service.isRunning());
+        assertEquals(baseUrl + basePath.substring(1), service.getUrl().toString());
+    }
+
+    @Test
+    public void checkAbilityToValidateBasePathForEmptyBasePath() {
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
+            service = new AppiumServiceBuilder().withArgument(BASEPATH, "").build();
```
Comment (valfirst) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:345]: no need to init `service` variable here

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 16]
```diff
@@ -312,4 +318,47 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
+        });
+        assertEquals(INVALID_BASE_PATH_ERROR_MESSAGE, exception.getMessage());
+    }
+
+    @Test
+    public void checkAbilityToValidateBasePathForBlankBasePath() {
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
+            service = new AppiumServiceBuilder().withArgument(BASEPATH, "   ").build();
```
Comment (valfirst) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:353]: no need to init `service` variable here

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 17]
```diff
@@ -312,4 +318,47 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
+        });
+        assertEquals(INVALID_BASE_PATH_ERROR_MESSAGE, exception.getMessage());
+    }
+
+    @Test
+    public void checkAbilityToValidateBasePathForNullBasePath() {
+        assertThrows(NullPointerException.class, () -> {
+            service = new AppiumServiceBuilder().withArgument(BASEPATH, null).build();
```
Comment (valfirst) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:361]: no need to init `service` variable here

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 12 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:53]: Import order got changed as per alphanumeric order after I formatted the code as per the check-style recommended, should I change it back?

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 17 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java]: Updated in the new commit, thanks for highlighting

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 16 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java]: Updated in the new commit, thanks for highlighting

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 15 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java]: Updated in the new commit, thanks for highlighting

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 14 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java]: Updated in the new commit, thanks for highlighting

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 12 above)
Comment (robinGupta11392) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: Updated in the new commit

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 11 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java]: Updated in the new commit, thanks for highlighting

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java [hunk 18]
```diff
@@ -312,4 +316,45 @@ public void checkAbilityToStartServiceWithLogFileUsingShortFlag() {
...
+        String basePath = "/wd/";
+        service = new AppiumServiceBuilder().withArgument(BASEPATH, basePath).build();
+        service.start();
+        assertTrue(service.isRunning());
+        assertEquals(baseUrl + basePath.substring(1), service.getUrl().toString());
+    }
+
+    @Test
+    public void checkAbilityToValidateBasePathForEmptyBasePath() {
+        assertThrows(IllegalArgumentException.class, () -> {
```
Comment (mykola-mokhnach) [inline src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:342]: FYI: you could add expected exception signature to Test annotations: https://www.guru99.com/junit-exception-test.html

**Review by mykola-mokhnach — APPROVED** [review]

File: src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java (hunk 18 above)
Comment (robinGupta11392) [reply src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java:342]: Noted with thanks.

**Review by valfirst — APPROVED** [review]
